src/external_trait_impls/rayon/raw.rs

```diff
@@ -0,0 +1,539 @@
+            &mut bucket.as_mut().1
+        });
+        folder.consume_iter(iter)
+    }
+}
+
+
+/// Parallel iterator over entries of a consumed map.
+///
+/// This iterator is created by the [`into_par_iter`] method on [`HashMap`]
+/// (provided by the [`IntoParallelIterator`] trait).
+/// See its documentation for more.
+///
+/// [`into_par_iter`]: /hashbrown/struct.HashMap.html#method.into_par_iter
+/// [`HashMap`]: /hashbrown/struct.HashMap.html
+/// [`IntoParallelIterator`]: https://docs.rs/rayon/1.0/rayon/iter/trait.IntoParallelIterator.html
+pub struct IntoParIter<K, V> {
+    table: RawTable<(K, V)>,
+}
+
+impl<K: Send, V: Send> IntoParallelIterator for RawTable<(K, V)> {
+    type Item = (K, V);
+    type Iter = IntoParIter<K, V>;
+
+    #[inline]
+    fn into_par_iter(self) -> Self::Iter {
+        IntoParIter { table: self }
+    }
+}
+
+impl<K: Send, V: Send> ParallelIterator for IntoParIter<K, V> {
+    type Item = (K, V);
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        let iter = unsafe { self.table.iter().iter };
+        let alloc = self.table.into_alloc();
+        let producer = IntoParIterProducer { iter, alloc };
+        plumbing::bridge_unindexed(producer, consumer)
+    }
+}
+
+impl<'a, K: fmt::Debug, V: fmt::Debug> fmt::Debug for IntoParIter<K, V> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let iter = unsafe { self.table.iter() };
+        let entries = iter.map(|bucket| unsafe {
+            let (ref k, ref v) = *bucket.as_ref();
+            (k, v)
+        });
+
+        f.debug_list().entries(entries).finish()
+    }
+}
+
+struct IntoParIterProducer<K, V> {
+    iter: RawIterRange<(K, V)>,
+    alloc: Option<(NonNull<u8>, Layout)>,
+}
+
+unsafe impl<K: Send, V: Send> Send for IntoParIterProducer<K, V> {}
+
+impl<K: Send, V: Send> UnindexedProducer for IntoParIterProducer<K, V> {
+    type Item = (K, V);
+
+    fn split(mut self) -> (Self, Option<Self>) {
+        let (left_iter, right_iter) = unsafe { self.iter.clone().split() };
+
+        self.iter = left_iter;
+        let right = right_iter.map(|iter| IntoParIterProducer { iter, alloc: None });
+
+        (self, right)
+    }
+
+    fn fold_with<F>(mut self, folder: F) -> F
+    where
+        F: Folder<Self::Item>,
+    {
+        let iter = self.iter.by_ref().map(|bucket| unsafe {
+            bucket.read()
+        });
+        folder.consume_iter(iter)
+    }
+}
+
+impl<K, V> Drop for IntoParIterProducer<K, V> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            // Drop all remaining elements
+            if mem::needs_drop::<(K, V)>() {
+                while let Some(item) = self.iter.next() {
+                    item.drop();
+                }
+            }
+
+            // Free the table
+            if let Some((ptr, layout)) = self.alloc {
+                dealloc(ptr.as_ptr(), layout);
+            }
```

Comment:
I don't think this is safe to `dealloc` from one of potentially many producers sharing this data.  You're keeping `alloc` in the left-most split, but nothing guarantees that will outlive the other splits.